### PR TITLE
fbfw-19

### PR DESCRIPTION
**Pull request by hyukcodazen**
added buttons in home page/my games dashboard, my games page and game header in game board.

**Only opt in users can see the buttons.**

Screen Shot 2022-01-05 at 9 43 08 AM

**check account settings page and opt in for beta features.**

Screen Shot 2022-01-05 at 9 43 40 AM

**play beta button is located in  My Games tab from Home, My Games page, and Game page.**

Screen Shot 2022-01-05 at 9 43 48 AM
Screen Shot 2022-01-05 at 9 43 54 AM
Screen Shot 2022-01-05 at 9 44 07 AM

### CODE REVIEW

File: gamepanel/game.php [hunk 1]
```diff
@@ -143,12 +143,16 @@ function gameTimeRemaining()
...
+					<span class="gameTimeRemainingNextPhase">'.l_t('Start:').'</span> '.$this->processTimetxt().' <span class="timestampGamesWrapper"> ('.libTime::detailedText($this->processTime).') </span>
+				';
 		else
 		{
-			$buf = '<span class="gameTimeRemainingNextPhase">'.l_t('Next:').'</span> '.
-				$this->processTimetxt().' ('.libTime::detailedText($this->processTime).')';
+			$buf =
+				'
+					<span class="gameTimeRemainingNextPhase">'.l_t('Next:').'</span> '.$this->processTimetxt().' <span class="timestampGamesWrapper"> ('.libTime::detailedText($this->processTime).') </span>
+				';
```
Comment (nebevets) [inline gamepanel/game.php:155]: it looks like 'Next' and 'Start' are the only things changing here. can we simplify this logic to have a ternary for the text that goes in the span?

File: gamepanel/game.php [hunk 2]
```diff
@@ -560,17 +576,25 @@ function joinBar()
...
+		// 	l_t($this->Members->isJoined()?'Open':'View').'</a>';
 
-		return '<form method="get" action="board.php#gamePanel"><div>
-			<input type="hidden" name="gameID" value="'.$this->id.'" />
-			<input type="submit" value="" class="form-submit" />
-			</div></form>';
+		// return '<form method="get" action="board.php#gamePanel"><div>
+		// 	<input type="hidden" name="gameID" value="'.$this->id.'" />
+		// 	<input type="submit" value="" class="form-submit" />
+		// 	</div></form>';
```
Comment (nebevets) [inline gamepanel/game.php:597]: usually we remove commented code.

File: gamepanel/game.php [hunk 3]
```diff
@@ -560,17 +576,25 @@ function joinBar()
 	function openBar()
 	{
 		global $User;
+		$playBeta = '';
+		if (($User->optInFeatures & 0x1) > 0) { $playBeta = '<a href="board.php?gameID='.$this->id.'/beta #gamePanel" style="margin-left: 40px">'.l_t('Play Beta').'</a> '; }
```
Comment (nebevets) [inline gamepanel/game.php:580]: i see that `$User->optInFeatures & 0x1) > 0` is being used in two different functions. can we dry this comparison up by assigning it to a variable?

File: gamepanel/game.php [hunk 4]
```diff
@@ -238,9 +253,10 @@ function titleBar()
 	{
 		$rightTop = '
 			<div class="titleBarRightSide">
-				<div>
-				<span class="gameTimeRemaining">'.$this->gameTimeRemaining().'</span></div>'.
-			'</div>';
+					<span class="gameTimeRemaining">'.$this->gameTimeRemaining().'</span>
+					<span class="gamePlayBeta">'.$this->gamePlayBeta().'</span>
+					<div style="clear:both"></div>
```
Comment (nebevets) [inline gamepanel/game.php:258]: this seems weird to me. why are we using a empty div to clear?

File: gamepanel/gamehome.php [hunk 5]
```diff
@@ -188,32 +189,35 @@ function links()
 		{
 			$noticesStatus = 5;
 			$SubmitName = 'Toggle Notices';
+			if (($User->optInFeatures & 0x1) > 0) { $playBeta = '<a href="board.php?gameID='.$this->id.'/beta #gamePanel" >'.l_t('Play Beta').'</a> '; }
```
Comment (nebevets) [inline gamepanel/gamehome.php:192]: again, it seems like maybe this should be a property of User? like isBeta or something so we don't have to keep making this comparison

File: gamepanel/gamehome.php [hunk 6]
```diff
@@ -188,32 +189,35 @@ function links()
...
-				return '<div class="bar homeGameLinks barAlt'.libHTML::alternate().'">
-				<form action="#" method="post">
-				'.libAuth::formTokenHTML().'
-					<a href="board.php?gameID='.$this->id.'">'.l_t('Open').'</a>
-					<input type="hidden" value="'.$this->id.'" name="gameToggleName" />
-					<input type="submit" title="Turn on/off the notifications for this game." style="float: right;" class = "home-submit toggle-notice" name="submit" value="'.$SubmitName.'"/>
-					</form>
+				return 
+					'<div class="bar homeGameLinks barAlt'.libHTML::alternate().'">
+						<form action="#" method="post" style="display: flex; justify-content: space-between;">
```
Comment (nebevets) [inline gamepanel/gamehome.php:201]: why not use a class instead of inline styles?

File: gamepanel/gamehome.php [hunk 7]
```diff
@@ -188,32 +189,35 @@ function links()
...
-					<form action="#" method="post">
-					'.libAuth::formTokenHTML().'
-					<a href="board.php?gameID='.$this->id.'#gamePanel">'.l_t('Open').'</a> 
-					<input type="hidden" value="'.$this->id.'" name="gameToggleName" />
-					<input type="submit" title="Turn on/off the notifications for this game." style="float: right;" class = "home-submit toggle-notice" name="submit" value="'.$SubmitName.'"/>
-					</form>
-					
+				return 
+					'<div class="bar homeGameLinks barAlt'.libHTML::alternate().'">
+						<form action="#" method="post" style="display: flex; justify-content: space-between;">
```
Comment (nebevets) [inline gamepanel/gamehome.php:214]: use the same class here instead of inline css?

**Review by nebevets — COMMENTED** [review]
looking good, just some questions and comments.

**Comment by nebevets** [comment]
Please add a test plan with screenshots that show what changes you made in action

File: gamepanel/game.php (hunk 4 above)
Comment (hyukcodazen) [reply gamepanel/game.php:258]: This is how they position the items on both desktop and mobile with float. Ideally, I'd go with flex but that's going to require changes to the entire tiltle bar.

File: gamepanel/gamehome.php (hunk 5 above)
Comment (hyukcodazen) [reply gamepanel/gamehome.php:192]: so they are using this integer bitset to capture if user is opted in to various experimental features. Looks like they did it this way to minimize the effort of adding or removing additional features without DB changing. I thought having different properties for each features would make more sense with having them treated separately in the DB. I'm not familiar with this concept so please let me know if I misinterpreted

Comment (bryancodazen) [reply gamepanel/gamehome.php:192]: @hyukcodazen maybe we can abstract this comparison to an easy to call ENUM of some sort. As Steve mentioned, even though this bit comparison works it is difficult to know exactly what it's doing without diving into it.

File: gamepanel/gamehome.php (hunk 6 above)
Comment (bryancodazen) [reply gamepanel/gamehome.php]: @hyukcodazen bump

File: gamepanel/gamehome.php (hunk 7 above)
Comment (bryancodazen) [reply gamepanel/gamehome.php]: @hyukcodazen bump

**Comment by hyukcodazen** [comment]
@bryancodazen Please let me know if I can merge this. Thanks!

**Comment by bryancodazen** [comment]
@hyukcodazen 

It seems when I click "Play Beta" I am redirected to the wrong URL: 
Screen Shot 2022-01-05 at 4 50 52 PM

What was the final consensus for this? I see the "Play Beta" button but when I click it I don't join the game. 
Screen Shot 2022-01-05 at 5 02 46 PM

In the mocks, this doesn't have two "Play Beta" buttons, it only has the bottom left one. 
Screen Shot 2022-01-05 at 4 55 37 PM

**Review by bryancodazen — CHANGES_REQUESTED** [review]
@hyukcodazen 

It seems when I click "Play Beta" I am redirected to the wrong URL: 
Screen Shot 2022-01-05 at 4 50 52 PM

What was the final consensus for this? I see the "Play Beta" button but when I click it I don't join the game. 
Screen Shot 2022-01-05 at 5 02 46 PM

In the mocks, this doesn't have two "Play Beta" buttons, it only has the bottom left one. 
Screen Shot 2022-01-05 at 4 55 37 PM

**Comment by hyukcodazen** [comment]
@bryancodazen 

I haven't got any updates for the 'Play Beta' button for joinable games. I guess I will just remove it from now and ask Jason tmrw if we are going to go with new modal or a button.

I did not capture 'play beta' button showing up on all the game cards. They all share same html so I need to find a way to split them.

For the URL part, it seems like you are redirected but map isn't showing up. Not sure how it's happening. Works find on my local. Can you check the regular 'Open' button?

**Comment by hyukcodazen** [comment]
url fix
beta button showing up twice from game panel fix
hide beta button on spectate
new paly beta button for joinable game

**Review by bryancodazen — APPROVED** [review]
Awesome! The only thing I would update (we can change this later) is the redirect should be to `http://localhost/beta?gameID=12` not `http://localhost/beta/?gameID=12`

**Pull request merged**